zebra-consensus/src/block/subsidy/funding_streams.rs

```diff
@@ -50,3 +48,114 @@ pub fn funding_stream_values(
+///
+/// [7.10]: https://zips.z.cash/protocol/protocol.pdf#fundingstreams
+fn funding_stream_address_period(height: Height, network: Network) -> u32 {
+    (height.0 + (POST_BLOSSOM_HALVING_INTERVAL.0) - (height_for_halving(network).0))
+        / (FUNDING_STREAM_ADDRESS_CHANGE_INTERVAL.0)
+}
+
+/// Returns the position in the address slice for each funding stream
+/// as described in [protocol specification §7.10][7.10]
+///
+/// [7.10]: https://zips.z.cash/protocol/protocol.pdf#fundingstreams
+fn funding_stream_address_index(height: Height, network: Network) -> usize {
+    let num_addresses = match network {
+        Network::Mainnet => FUNDING_STREAMS_N_ADDRESSES_MAINNET,
+        Network::Testnet => FUNDING_STREAMS_N_ADDRESSES_TESTNET,
+    };
+
+    let index = 1u32
+        .checked_add(funding_stream_address_period(height, network))
+        .unwrap()
+        .checked_sub(funding_stream_address_period(
+            FUNDING_STREAM_HEIGHT_RANGES.get(&network).unwrap().start,
+            network,
+        ))
+        .unwrap() as usize;
+    assert!(index > 0 && index <= num_addresses);
+
+    index
+}
+
+/// Return the address corresponding to this height for this funding stream receiver.
+pub fn funding_stream_address(
+    height: Height,
+    network: Network,
+    receiver: FundingStreamReceiver,
+) -> Address {
+    let index = funding_stream_address_index(height, network) - 1;
+
+    let address = match receiver {
+        FundingStreamReceiver::Ecc => match network {
+            Network::Mainnet => FUNDING_STREAM_ECC_ADDRESSES_MAINNET[index].to_string(),
+            Network::Testnet => FUNDING_STREAM_ECC_ADDRESSES_TESTNET[index].to_string(),
+        },
+        FundingStreamReceiver::ZcashFoundation => match network {
+            Network::Mainnet => FUNDING_STREAM_ZF_ADDRESSES_MAINNET[index].to_string(),
+            Network::Testnet => FUNDING_STREAM_ZF_ADDRESSES_TESTNET[index].to_string(),
+        },
+        FundingStreamReceiver::MajorGrants => match network {
+            Network::Mainnet => FUNDING_STREAM_MG_ADDRESSES_MAINNET[index].to_string(),
+            Network::Testnet => FUNDING_STREAM_MG_ADDRESSES_TESTNET[index].to_string(),
+        },
+    };
+    Address::from_str(&address).expect("Address should deserialize")
+}
+
+/// Given a founders reward address, create a script and check if it is the same
+/// as the given lock_script as described in [protocol specification §7.10][7.10]
+///
+/// [7.10]: https://zips.z.cash/protocol/protocol.pdf#fundingstreams.
+pub fn check_script_form(lock_script: Script, address: Address) -> bool {
+    let mut address_hash = address
+        .zcash_serialize_to_vec()
+        .expect("we should get address bytes here");
+
+    address_hash = address_hash[2..22].to_vec();
+    address_hash.insert(0, OpCode::Push20Bytes as u8);
+    address_hash.insert(0, OpCode::Hash160 as u8);
+    address_hash.insert(address_hash.len(), OpCode::Equal as u8);
```

Comment:
From https://github.com/oxarbitrage/zebra/pull/160#pullrequestreview-800589185

> It feels a bit weird to be doing this in Rust, when the rest of our script validation happens in `zcash_script`, which calls `zcashd`'s C++ implementation. But adding another `unsafe` C++ function could be complicated, and it's more risky. (And this function might not even exist in `zcashd`.)
>
> So I think this code is simple enough to re-implement. We just need to test it carefully.
>
> Are there any `zcashd` test vectors we could use?

Reply:
I dont think zcashd has test vectors for this. I just synced the mainnet up to tip using the last version of this pull request. that gives us some coverage that all blocks since Canopy until now had been validated properly.